brainstorming.md

```diff
@@ -0,0 +1,372 @@
+
+- a table schema, expressed as an `Arrow.RecordBatch` type, defining the number of columns, and their respective names and types
+- the _primary_ index column name
+- the _secondary_ index column name
+
+A SOMAMatrix must contain (minimally) two columns which are the primary and secondary index columns. The index columns _must_ be of `string` type. Index values are non-null and must be unique (ie, no duplicates are allowed) within the object. The index is sparse and unordered, ie, supported indexing is by ID value only, with no indexing order or positional (offset-based) indexing.
+
+> ⚠️ **Issue** - SOMAMatrix and SOMADataFrame differ only in the number of dimensions (2 vs 1 respectively). Should they be unified into a single type object/API of more general nature? Clarity vs flexibility?
+
+## SOMA
+
+The SOMA type is a specialized SOMACollection, representing an annotated 2-D matrix of observations across a set of variables. The SOMA composes SOMACollection, SOMADataFrame and SOMAMatrix types using a set of predefined fields (keys & values), to represent the annotated matrix. Each predefined field in the SOMA has a well-defined type, dimensionality and indexing convention. Other user defined data may be added to a SOMA, as a SOMA is a specialization of a SOMACollection.
+
+Implementations _should_ enforce the naming and indexing constraints on these pre-defined fields. Pre-defined fields are distinguished from other user-defined collection elements, where no schema or indexing semantics are presumed or enforced.
+
+The predefined fields within SOMA share common indices and `shape`:
+
+- obs_id - the observation, a `string` type
+- var_id - the variable, a `string` type
+
+SOMA pre-defined fields are principally differentiated by:
+
+- dimensionality, eg, 1-D (SOMADataFrame) or 2-D (SOMAMatrix)
+- index column names, `obs_id` or `var_id` or both
+
+The pre-defined fields are:
+
+| Field name | Field type                              | Field description                                                                                                                                                                                                                                               |
+| ---------- | --------------------------------------- | --------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
+| `obs`      | `SOMADataFrame`                         | Primary annotations on the _observation_ axis. Must be of length #observations (ie, every observation _must_ be defined in this field), and must contain and be indexed by the `obs_id` column. The contents of `obs_id` define the _observation_ index domain. |
```

Comment:
i really like how specific you are here. right here is where that's _most_ needed.